pkg/blockbuilder/blockbuilder.go

```diff
@@ -274,14 +274,65 @@ func (b *BlockBuilder) nextConsumeCycle(ctx context.Context, cycleEnd time.Time)
 		return nil
 	}
 
-	// TODO(v): rebalancing can happen between the calls to consumePartitions; if that happens, the instance may loose
+	// TODO(v): rebalancing can happen between the calls to consumePartition; if that happens, the instance may loose
 	// the ownership of some of its partitions
 	// TODO(v): test for this case
+
+	offsets, err := kadm.NewClient(b.kafkaClient).FetchOffsetsForTopics(ctx, b.cfg.Kafka.Topic)
+	if err != nil {
+		return fmt.Errorf("fetch offsets for topics: %w", err)
+	}
 	for part, lag := range parts {
-		err := b.consumePartitions(ctx, cycleEnd, part, lag)
-		if err != nil {
-			level.Error(b.logger).Log("msg", "failed to consume partition", "err", err, "part", part)
+		// We look at the commit offset timestamp to determine how far behind we are lagging
+		// relative to the cycleEnd. We consume the partition in parts accordingly.
+		offset, ok := offsets.Lookup(b.cfg.Kafka.Topic, part)
+		var commitRecTs, seenTillTs, lastBlockEnd int64
+		var commitRecTime time.Time
+		if ok {
+			commitRecTs, seenTillTs, lastBlockEnd, err = unmarshallCommitMeta(offset.Metadata)
```

Comment:
Even after fixing the actual commit (#8354), this still doesn't work as we expected it. The `offset.Metadata` here, for some reason, contains the group member-id — not the metadata we stored on commit.

I feel this is **a bug** `kadm` but I'll let you delve into it, if you don't mind. Please, let me know if you stuck there.

---

To start debugging it, this is what the `offsets` response looks like

```
part=1
offsets="map[mimir-ingest:map[
  0:{Offset:{Topic:mimir-ingest Partition:0 At:0 LeaderEpoch:-1 Metadata:} Err:<nil>}
  1:{Offset:{Topic:mimir-ingest Partition:1 At:26815 LeaderEpoch:2 Metadata:block-builder-1-0889164f-7889-4714-a11e-f2f16bc3d34e} Err:<nil>}
  2:{Offset:{Topic:mimir-ingest Partition:2 At:26836 LeaderEpoch:2 Metadata:block-builder-1-0889164f-7889-4714-a11e-f2f16bc3d34e} Err:<nil>}
  3:{Offset:{Topic:mimir-ingest Partition:3 At:0 LeaderEpoch:-1 Metadata:} Err:<nil>}
  4:{Offset:{Topic:mimir-ingest Partition:4 At:0 LeaderEpoch:-1 Metadata:} Err:<nil>}
  5:{Offset:{Topic:mimir-ingest Partition:5 At:0 LeaderEpoch:-1 Metadata:} Err:<nil>}
  6:{Offset:{Topic:mimir-ingest Partition:6 At:0 LeaderEpoch:-1 Metadata:} Err:<nil>}
  7:{Offset:{Topic:mimir-ingest Partition:7 At:0 LeaderEpoch:-1 Metadata:} Err:<nil>}
  8:{Offset:{Topic:mimir-ingest Partition:8 At:0 LeaderEpoch:-1 Metadata:} Err:<nil>}
  9:{Offset:{Topic:mimir-ingest Partition:9 At:0 LeaderEpoch:-1 Metadata:} Err:<nil>}
]]"
```

Note the Metadata field for partitions 1 and 2.